scipy/stats/stats.py

```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
     >>> stats.pearsonr([1,2,3,4,5], [5,6,7,8,7])
     (0.83205029433784372, 0.080509573298498519)
     """
-    # x and y should have same length.
+    n = len(x)
+    if n != len(y):
+        raise ValueError('x and y must have the same length.')
+
+    if n < 2:
+        raise ValueError('x and y must have length at least 2.')
+
     x = np.asarray(x)
     y = np.asarray(y)
-    n = len(x)
-    mx = x.mean()
-    my = y.mean()
-    xm, ym = x - mx, y - my
-    r_num = np.add.reduce(xm * ym)
-    r_den = np.sqrt(_sum_of_squares(xm) * _sum_of_squares(ym))
-    r = r_num / r_den
+
+    # If an input is constant, the correlation coefficient is not defined.
+    if (x == x[0]).all() or (y == y[0]).all():
+        warnings.warn(PearsonRConstantInputWarning())
+        return np.nan, np.nan
+
+    if n == 2:
+        return np.float64(np.sign(x[1] - x[0])*np.sign(y[1] - y[0])), 1.0
+
+    xmean = x.mean(dtype=np.float64)
+    ymean = y.mean(dtype=np.float64)
+
+    xm = x - xmean
+    ym = y - ymean
+
+    # Unlike np.linalg.norm or the expression sqrt((xm*xm).sum()),
+    # scipy.linalg.norm(xm) does not overflow if xm is, for example,
+    # [-5e210, 5e210, 3e200, -3e200]
+    normxm = linalg.norm(xm)
+    normym = linalg.norm(ym)
+
+    threshold = 1e-13
+    if normxm < threshold*abs(xmean) or normym < threshold*abs(ymean):
+        # If all the values in x (likewise y) are very close to the mean,
+        # the loss of precision that occurs in the subtraction xm = x - xmean
+        # might result in large errors in r.
+        warnings.warn(PearsonRNearConstantInputWarning())
+
+    r = np.dot(xm/normxm, ym/normym)
 
     # Presumably, if abs(r) > 1, then it is only some small artifact of
     # floating point arithmetic.
     r = max(min(r, 1.0), -1.0)
```

Comment:
in one of the issues, this line caused a problem when r = inf or -inf

Reply:
That's https://github.com/scipy/scipy/issues/9353.  In that issue, the values in `y` are so small that in the calculation `sum(y**2)`, the squared terms underflow to 0, which caused the denominator in the calculation of `r` to be 0.  That problem is fixed by the reformulation in this pull request, and the data from that issue is used in one of the unit tests.